File: client/go/pico.go
package picomq

import (
	"context"
	"encoding/json"
	"net/http"
	"net/url"
	"strconv"
	"time"
)

func (c *PicoClient) Stream(name string) *PicoStream { return &PicoStream{client: c, name: name} }

func (c *PicoClient) Create(ctx context.Context, name, contentType string, ttl time.Duration) (created bool, err error) {
	err = c.core.run(ctx, func() error {
		headers := make(http.Header)
		headers.Set("Content-Type", contentType)
		if ttl > 0 {
			headers.Set("Pico-TTL", strconv.FormatInt(int64(ttl/time.Second), 10))
		}
		response, callErr := c.core.send(ctx, wireRequest{method: http.MethodPut, url: c.core.streamURL(name, nil), headers: headers})
		if callErr != nil {
			return callErr
		}
		created = response.StatusCode == http.StatusCreated
		_, callErr = expectPico(response, http.StatusOK, http.StatusCreated)
		return callErr
	})
	return
}

func (c *PicoClient) Head(ctx context.Context, name string) (info *StreamInfo, err error) {
	err = c.core.run(ctx, func() error {
		response, callErr := c.core.send(ctx, wireRequest{method: http.MethodHead, url: c.core.streamURL(name, nil)})
		if callErr != nil {
			return callErr
		}
		if response.StatusCode == http.StatusNotFound {
			response.Body.Close()
			info = nil
			return nil
		}
		if _, callErr = expectPico(response, http.StatusOK); callErr != nil {
			return callErr
		}
		info = &StreamInfo{Name: name, ContentType: response.Header.Get("Content-Type"), Start: defaultString(response.Header.Get("Pico-Start-Seq"), "0"), Next: defaultString(response.Header.Get("Pico-Next-Seq"), "0"), Closed: headerBool(response.Header, "Pico-Closed"), TTL: time.Duration(headerUint(response.Header, "Pico-TTL")) * time.Second, ExpiresAt: response.Header.Get("Pico-Expires-At")}
		return nil
	})
	return
}

func (c *PicoClient) Append(ctx context.Context, name string, records []AppendRecord) (AppendAck, error) {
	return c.append(ctx, name, records, nil)
}

func (c *PicoClient) AppendAs(ctx context.Context, name string, records []AppendRecord, producer ProducerRef) (ProducerAck, error) {
	headers := make(http.Header)
	headers.Set("Pico-Producer-Id", producer.ID)
	headers.Set("Pico-Producer-Epoch", strconv.FormatUint(producer.Epoch, 10))
	headers.Set("Pico-Producer-Seq", strconv.FormatUint(producer.Seq, 10))
	ack, responseHeaders, err := c.appendWithHeaders(ctx, name, records, headers)
	if err != nil {
		return ProducerAck{}, err
	}
	applied := responseHeaders.Get("Pico-Start-Seq") != ""
	return ProducerAck{Applied: applied, Duplicate: !applied && len(records) > 0, Ack: ack}, nil
}

func (c *PicoClient) append(ctx context.Context, name string, records []AppendRecord, headers http.Header) (AppendAck, error) {
	ack, _, err := c.appendWithHeaders(ctx, name, records, headers)
	return ack, err
}

func (c *PicoClient) appendWithHeaders(ctx context.Context, name string, records []AppendRecord, headers http.Header) (AppendAck, http.Header, error) {
	payload, err := encodeBatch(records)
	if err != nil {
		return AppendAck{}, nil, &ClientError{Kind: ErrorBadRequest, Code: "invalid_record", Message: err.Error(), Cause: err}
	}
	if headers == nil {
		headers = make(http.Header)
	}
	headers.Set("Content-Type", "application/vnd.picomq.batch")
	response, err := c.core.send(ctx, wireRequest{method: http.MethodPost, url: c.core.streamURL(name, nil), headers: headers, body: payload})
	if err != nil {
		return AppendAck{}, nil, err
	}
	if _, err = expectPico(response, http.StatusOK); err != nil {
		return AppendAck{}, nil, err
	}
	next := defaultString(response.Header.Get("Pico-Next-Seq"), "0")
	return AppendAck{Start: defaultString(response.Header.Get("Pico-Start-Seq"), next), Next: next, Timestamp: int64(headerUint(response.Header, "Pico-Timestamp"))}, response.Header, nil
}

func (c *PicoClient) Read(ctx context.Context, name, from string, options ReadOptions) (page ReadPage, err error) {
	err = c.core.run(ctx, func() error {
		query := url.Values{"format": {"binary"}, "seq": {from}}
		if options.Limits.Count > 0 {
			query.Set("count", strconv.FormatUint(options.Limits.Count, 10))
		}
		if options.Limits.Bytes > 0 {
			query.Set("bytes", strconv.FormatUint(options.Limits.Bytes, 10))
		}
		if options.Live == LiveLongPoll {
			query.Set("live", "long-poll")
		}
		response, callErr := c.core.send(ctx, wireRequest{method: http.MethodGet, url: c.core.streamURL(name, query)})
		if callErr != nil {
			return callErr
		}
		expected := []int{http.StatusOK}
		if options.Live == LiveLongPoll {
			expected = append(expected, http.StatusNoContent)
		}
		data, callErr := expectPico(response, expected...)
		if callErr != nil {
			return callErr
		}
		page = ReadPage{Next: defaultString(response.Header.Get("Pico-Next-Seq"), from), UpToDate: headerBool(response.Header, "Pico-Up-To-Date") || response.StatusCode == http.StatusNoContent, Closed: headerBool(response.Header, "Pico-Closed")}
		if len(data) > 0 {
			page.Records, callErr = decodeBatch(data)
			if callErr != nil {
				return invalidResponse(callErr)
			}
		}
		return nil
	})
	return
}

func (c *PicoClient) List(ctx context.Context, prefix string, limit uint64) (listing StreamListing, err error) {
	err = c.core.run(ctx, func() error {
		query := url.Values{"prefix": {prefix}}
		if limit > 0 {
			query.Set("limit", strconv.FormatUint(limit, 10))
		}
		response, callErr := c.core.send(ctx, wireRequest{method: http.MethodGet, url: c.core.streamURL("/", query)})
		if callErr != nil {
			return callErr
		}
		data, callErr := expectPico(response, http.StatusOK)
		if callErr != nil {
			return callErr
		}
		// The server uses snake_case; decode explicitly to keep the public model independent.
		var raw struct {
			Streams []struct {
				Name        string `json:"name"`
				ContentType string `json:"content_type"`
				Start       uint64 `json:"start_seq"`
				Next        uint64 `json:"next_seq"`
				Closed      bool   `json:"closed"`
				TTL         uint64 `json:"ttl"`
				ExpiresAt   string `json:"expires_at"`
			} `json:"streams"`
			HasMore bool `json:"has_more"`
		}
		if callErr = json.Unmarshal(data, &raw); callErr != nil {
			return invalidResponse(callErr)
		}
		listing = StreamListing{Streams: make([]StreamInfo, 0, len(raw.Streams))}
		listing.HasMore = raw.HasMore
		for _, item := range raw.Streams {
			listing.Streams = append(listing.Streams, StreamInfo{Name: item.Name, ContentType: item.ContentType, Start: strconv.FormatUint(item.Start, 10), Next: strconv.FormatUint(item.Next, 10), Closed: item.Closed, TTL: time.Duration(item.TTL) * time.Second, ExpiresAt: item.ExpiresAt})
		}
		return nil
	})
	return
}

func (c *PicoClient) Trim(ctx context.Context, name string, seq uint64) (start string, err error) {
	err = c.core.run(ctx, func() error {
		headers := make(http.Header)
		headers.Set("Pico-Trim-Seq", strconv.FormatUint(seq, 10))
		response, callErr := c.core.send(ctx, wireRequest{method: http.MethodPost, url: c.core.streamURL(name, nil), headers: headers})
		if callErr != nil {
			return callErr
		}
		_, callErr = expectPico(response, http.StatusOK)
		start = defaultString(response.Header.Get("Pico-Start-Seq"), "0")
		return callErr
	})
	return
}

func (c *PicoClient) Close(ctx context.Context, name string) (next string, err error) {
	err = c.core.run(ctx, func() error {
		headers := make(http.Header)
		headers.Set("Pico-Closed", "true")
		response, e := c.core.send(ctx, wireRequest{method: http.MethodPost, url: c.core.streamURL(name, nil), headers: headers})
		if e != nil {
			return e
		}
		_, e = expectPico(response, http.StatusOK)
		next = defaultString(response.Header.Get("Pico-Next-Seq"), "0")
		return e
	})
	return
}

func (c *PicoClient) Delete(ctx context.Context, name string) (deleted bool, err error) {
	err = c.core.run(ctx, func() error {
		response, e := c.core.send(ctx, wireRequest{method: http.MethodDelete, url: c.core.streamURL(name, nil)})
		if e != nil {
			return e
		}
		if response.StatusCode == http.StatusNotFound {
			response.Body.Close()
			deleted = false
			return nil
		}
		_, e = expectPico(response, http.StatusNoContent)
		deleted = e == nil
		return e
	})
	return
}

func defaultString(value, fallback string) string {
	if value == "" {
		return fallback
	}
	return value
}

File: picomq/pico-client/src/pico.rs
use async_trait::async_trait;
use bytes::Bytes;
use picomq_protocol::WireRequest;
use picomq_protocol::pico::{
    AppendRequest, AppendResponse, CloseRequest, CloseResponse, CreateRequest, CreateResponse,
    DeleteRequest, DeleteResponse, HeadRequest, HeadResponse, LIVE_LONG_POLL, ListRequest, Listing,
    ReadRequest, ReadResponse, SEQ_BEGINNING, TrimRequest, TrimResponse, decode_error,
};
use picomq_protocol::record::PicoRecord;
use reqwest::Response;

use crate::error::{ClientError, ErrorKind, Result};
use crate::retry::RetryPolicy;
use crate::types::{
    AppendAck, Live, Protocol, ReadLimits, ReadPage, Record, StreamApi, StreamInfo, StreamListing,
};

pub use picomq_protocol::pico::Producer as ProducerRef;

#[derive(Debug, Clone)]
pub struct ProducerAck {
    pub applied: bool,
    pub duplicate: bool,
    pub ack: AppendAck,
}

#[derive(Clone)]
pub struct PicoClient {
    http: reqwest::Client,
    base_url: String,
    retry: RetryPolicy,
}

impl PicoClient {
    pub fn new(endpoint: &str) -> Result<Self> {
        Ok(Self::with_http(
            endpoint,
            default_http()?,
            RetryPolicy::none(),
        ))
    }

    pub fn with_http(endpoint: &str, http: reqwest::Client, retry: RetryPolicy) -> Self {
        Self {
            http,
            base_url: endpoint.trim_end_matches('/').to_owned(),
            retry,
        }
    }

    pub async fn append_as(
        &self,
        name: &str,
        records: &[Bytes],
        producer: &ProducerRef<'_>,
    ) -> Result<ProducerAck> {
        let records_wire = wire_records(records);
        let mut request = AppendRequest::new(name, &records_wire);
        request.producer = Some(*producer);
        let response = self.call(request.encode()).await?;
        let ack = AppendResponse::decode(response.headers());
        let next = seq_string(ack.next_seq);
        let applied = ack.start_seq.is_some();
        Ok(ProducerAck {
            applied,
            duplicate: !applied && !records.is_empty(),
            ack: AppendAck {
                start: ack
                    .start_seq
                    .map(|v| v.to_string())
                    .unwrap_or_else(|| next.clone()),
                next,
                timestamp: ack.timestamp,
            },
        })
    }

    pub async fn trim(&self, name: &str, seq: u64) -> Result<String> {
        let response = self
            .call(TrimRequest { stream: name, seq }.encode())
            .await?;
        Ok(seq_string(
            TrimResponse::decode(response.headers()).start_seq,
        ))
    }

    async fn call(&self, wire: WireRequest) -> Result<Response> {
        let ok = wire.ok;
        let response = send(&self.http, build(&self.http, &self.base_url, wire)).await?;
        expect(response, ok).await
    }

    async fn head_once(&self, name: &str) -> Result<Option<StreamInfo>> {
        let response = self.call(HeadRequest { stream: name }.encode()).await?;
        let Some(head) = HeadResponse::decode(response.status().as_u16(), response.headers())
        else {
            return Ok(None);
        };
        Ok(Some(StreamInfo {
            name: name.to_owned(),
            content_type: head.content_type,
            start: seq_string(head.start_seq),
            next: seq_string(head.next_seq),
            closed: head.closed,
            ttl_seconds: head.ttl_seconds,
            expires_at: head.expires_at,
            retention_ms: head.retention_ms,
        }))
    }

    async fn read_once(
        &self,
        name: &str,
        from: &str,
        live: Live,
        limits: ReadLimits,
    ) -> Result<ReadPage> {
        let mut request = ReadRequest::new(name, from);
        request.count = limits.count;
        request.bytes = limits.bytes;
        request.live = (live == Live::LongPoll).then_some(LIVE_LONG_POLL);
        let response = self.call(request.encode()).await?;

        let status = response.status().as_u16();
        let headers = response.headers().clone();
        let body = response.bytes().await?;
        let read = ReadResponse::decode(status, &headers, &body).map_err(invalid_response)?;

        Ok(ReadPage {
            up_to_date: read.up_to_date || (read.no_content && live == Live::LongPoll),
            records: read
                .records
                .into_iter()
                .map(|record| Record {
                    position: record.seq.to_string(),
                    timestamp: Some(record.record.timestamp),
                    key: record.record.key,
                    headers: record.record.headers,
                    body: record.record.body,
                })
                .collect(),
            next: read
                .next_seq
                .map(|v| v.to_string())
                .unwrap_or_else(|| from.to_owned()),
            closed: read.closed,
        })
    }

    async fn list_once(&self, prefix: &str, limit: u64) -> Result<StreamListing> {
        let request = ListRequest {
            prefix,
            limit,
            start_after: None,
        };
        let response = self.call(request.encode()).await?;
        let body = response.bytes().await?;
        let listing = Listing::decode(&body).map_err(invalid_response)?;

        Ok(StreamListing {
            streams: listing
                .streams
                .into_iter()
                .map(|entry| StreamInfo {
                    name: entry.name,
                    content_type: entry.content_type,
                    start: entry.start_seq.to_string(),
                    next: entry.next_seq.to_string(),
                    closed: entry.closed,
                    ttl_seconds: entry.ttl_seconds,
                    expires_at: entry.expires_at,
                    retention_ms: entry.retention_ms,
                })
                .collect(),
            has_more: listing.has_more,
        })
    }
}

#[async_trait]
impl StreamApi for PicoClient {
    fn protocol(&self) -> Protocol {
        Protocol::Pico
    }

    fn beginning(&self) -> String {
        SEQ_BEGINNING.to_owned()
    }

    fn now(&self) -> Result<String> {
        Err(ClientError::unsupported(
            "the Pico protocol has no `now` token; read from the stream's next seq",
        ))
    }

    async fn create(
        &self,
        name: &str,
        content_type: &str,
        ttl_seconds: Option<u64>,
    ) -> Result<bool> {
        let mut request = CreateRequest::new(name, content_type);
        request.ttl_seconds = ttl_seconds;
        let response = self.call(request.encode()).await?;
        Ok(CreateResponse::decode(response.status().as_u16(), response.headers()).created)
    }

    async fn head(&self, name: &str) -> Result<Option<StreamInfo>> {
        self.retry.run(|| self.head_once(name)).await
    }

    async fn append(
        &self,
        name: &str,
        records: &[Bytes],
        _content_type: &str,
    ) -> Result<AppendAck> {
        let records_wire = wire_records(records);
        let response = self
            .call(AppendRequest::new(name, &records_wire).encode())
            .await?;
        let ack = AppendResponse::decode(response.headers());
        let next = seq_string(ack.next_seq);
        Ok(AppendAck {
            start: ack
                .start_seq
                .map(|v| v.to_string())
                .unwrap_or_else(|| next.clone()),
            next,
            timestamp: ack.timestamp,
        })
    }

    async fn read(
        &self,
        name: &str,
        from: &str,
        live: Live,
        limits: ReadLimits,
    ) -> Result<ReadPage> {
        self.retry
            .run(|| self.read_once(name, from, live, limits))
            .await
    }

    async fn list(&self, prefix: &str, limit: u64) -> Result<StreamListing> {
        self.retry.run(|| self.list_once(prefix, limit)).await
    }

    async fn close(&self, name: &str) -> Result<String> {
        let response = self.call(CloseRequest { stream: name }.encode()).await?;
        Ok(seq_string(
            CloseResponse::decode(response.headers()).next_seq,
        ))
    }

    async fn delete(&self, name: &str) -> Result<bool> {
        let response = self.call(DeleteRequest { stream: name }.encode()).await?;
        Ok(DeleteResponse::decode(response.status().as_u16()).found)
    }
}

fn wire_records(records: &[Bytes]) -> Vec<PicoRecord> {
    records
        .iter()
        .map(|body| PicoRecord::new(body.clone()))
        .collect()
}

fn seq_string(seq: Option<u64>) -> String {
    seq.map(|v| v.to_string())
        .unwrap_or_else(|| SEQ_BEGINNING.to_owned())
}

fn invalid_response(e: picomq_protocol::CodecError) -> ClientError {
    ClientError::new(0, ErrorKind::Other, "invalid_response").with_message(Some(e.to_string()))
}

pub(crate) fn default_http() -> Result<reqwest::Client> {
    crate::http_client(&crate::ClientConfig::default())
}

pub(crate) fn build(
    http: &reqwest::Client,
    base_url: &str,
    wire: WireRequest,
) -> reqwest::RequestBuilder {
    let mut builder = http.request(wire.method, format!("{base_url}{}", wire.path_and_query));
    for (name, value) in wire.headers {
        builder = builder.header(name, value);
    }
    if !wire.body.is_empty() {
        builder = builder.body(wire.body);
    }
    builder
}

const MAX_REDIRECT_HOPS: usize = 5;

// Re-issues the request at the Location so every header, including
// Authorization, rides each redirect hop.
pub(crate) async fn send(
    http: &reqwest::Client,
    builder: reqwest::RequestBuilder,
) -> Result<Response> {
    let mut request = builder.build()?;
    for _ in 0..MAX_REDIRECT_HOPS {
        let base = request.url().clone();
        let retry = request.try_clone();
        let response = http.execute(request).await?;
        let status = response.status().as_u16();
        if status != 307 && status != 308 {
            return Ok(response);
        }
        let Some(target) = header(&response, "location").and_then(|loc| base.join(&loc).ok())
        else {
            return Err(ClientError::new(
                status,
                ErrorKind::Other,
                "redirect_without_location",
            ));
        };
        let mut next = retry.expect("bodies are Bytes, always clonable");
        *next.url_mut() = target;
        request = next;
    }
    Err(ClientError::new(0, ErrorKind::Other, "too_many_redirects"))
}

pub(crate) fn header(response: &Response, name: &str) -> Option<String> {
    response
        .headers()
        .get(name)
        .and_then(|value| value.to_str().ok())
        .map(str::to_owned)
}

pub(crate) async fn expect(response: Response, expected: &[u16]) -> Result<Response> {
    let status = response.status().as_u16();
    if expected.contains(&status) {
        return Ok(response);
    }
    let headers = response.headers().clone();
    let body = response.text().await.unwrap_or_default();
    Err(decode_error(status, &headers, &body).into())
}
